cdap-app-templates/cdap-etl/cdap-etl-batch-app/src/main/java/co/cask/cdap/app/etl/batch/EmailAction.java

```diff
@@ -0,0 +1,141 @@
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(properties.get(MESSAGE)),
+                                String.format("You must set the \'%s\' property to send an email.", MESSAGE));
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(properties.get(SUBJECT)),
+                                String.format("You must set the \'%s\' property to send an email.", SUBJECT));
+    Preconditions.checkArgument(!(Strings.isNullOrEmpty(properties.get(USERNAME)) ^
+                                  Strings.isNullOrEmpty(properties.get(PASSWORD))),
+                                String.format("You must either set both username and password " +
+                                                "or neither username nor password. " +
+                                                "Currently, they are username: \'%s\', and password: \'%s\'",
+                                              properties.get(USERNAME), properties.get(PASSWORD)));
+    if (Strings.isNullOrEmpty(properties.get(HOST))) {
+      properties.put(HOST, DEFAULT_HOST);
+    }
+    if (Strings.isNullOrEmpty(properties.get(PORT))) {
+      properties.put(PORT, Integer.toString(DEFAULT_PORT));
+    }
+    return WorkflowActionSpecification.Builder.with().setName(NAME).setDescription("")
+      .withOptions(properties).build();
+  }
+
+  @Override
+  public void run() {
+    properties = ((WorkflowActionNode) getContext().getWorkflowSpecification().getNodeIdMap()
+      .get(NAME)).getActionSpecification().getProperties();
+
+    Properties javaMailProperties = new Properties();
+    javaMailProperties.put("mail.smtp.host", !Strings.isNullOrEmpty(properties.get(HOST)) ?
+      properties.get(HOST) : DEFAULT_HOST);
+    javaMailProperties.put("mail.smtp.port", !Strings.isNullOrEmpty(properties.get(PORT)) ?
+      Integer.parseInt(properties.get(PORT)) : DEFAULT_PORT);
+    if (!(Strings.isNullOrEmpty(properties.get(USERNAME)))) {
+      javaMailProperties.put("mail.smtp.auth", true);
+    }
+    if ("SMTPS".equalsIgnoreCase(properties.get(PROTOCOL))) {
+      javaMailProperties.put("mail.smtp.ssl.enable", true);
+    }
+    if ("TLS".equalsIgnoreCase(properties.get(PROTOCOL))) {
+      javaMailProperties.put("mail.smtp.starttls.enable", true);
+    }
+
+    Authenticator authenticator = null;
+    if (!(Strings.isNullOrEmpty(properties.get(USERNAME)))) {
+      javaMailProperties.put("mail.smtp.auth", true);
+      authenticator = new Authenticator() {
+        @Override
+        public PasswordAuthentication getPasswordAuthentication() {
+          return new PasswordAuthentication(properties.get(USERNAME), properties.get(PASSWORD));
+        }
+      };
+    }
+
+    Session session = Session.getInstance(javaMailProperties, authenticator);
+    session.setDebug(true);
+    try {
+      Message msg = new MimeMessage(session);
+      msg.setFrom(new InternetAddress(FROM_ADDRESS));
```

Comment:
Should message have access to the tokens. the information from tokens should be part of the body of email.